### PR TITLE
Move Roslyn to .NET Core 2.0 SDK

### PR DESCRIPTION
**Pull request by jaredpar**
The core of this change is moving Roslyn to the 2.0 SDK. This required a bit of reworking of a number of projects as we were relying on bugs / undefined behavior in the old SDK. A quick summary of the changes:

- Many test projects moved to net461. This is just fallout from 2.0 SDK correctly enforcing target mapping. 
- Need to bootstrap CLI as a part of our core build. The 2.0 SDK is delivered via the CLI hence this is simply required.
- Remove RoslynRuntimeIdentifier as it's no longer necessary

### CODE REVIEW

File: build/Targets/Roslyn.Toolsets.Xunit.targets [hunk 1]
```diff
@@ -12,10 +12,39 @@
     <!-- xUnit.net WPF Runner -->
     <XUnitWpfPath>$(NuGetPackageRoot)\xunit.runner.wpf\$(xunitrunnerwpfVersion)\tools\xunit.runner.wpf.exe</XUnitWpfPath>
     <XUnitWpfArguments>"$(OutDir)\$(AssemblyName).dll"</XUnitWpfArguments>
-  </PropertyGroup>
 
-  <ItemGroup>
-    <PackageReference Include="xunit.runner.wpf" Version="$(xunitrunnerwpfVersion)" PrivateAssets="all" />
```
Comment (DustinCampbell) [inline build/Targets/Roslyn.Toolsets.Xunit.targets]: Does this mean we acquire xunit.runner.wpf any longer? Is there something that needs to happen to bring that back.

File: build/ToolsetPackages/BaseToolset.csproj [hunk 2]
```diff
@@ -27,5 +27,6 @@
     <PackageReference Include="RoslynTools.ReferenceAssemblies" Version="$(RoslynToolsReferenceAssembliesVersion)" />
     <PackageReference Include="GitLink" Version="$(GitLinkVersion)" />
     <PackageReference Include="Microsoft.VSSDK.BuildTools" Version="$(MicrosoftVSSDKBuildToolsVersion)" />
+    <PackageReference Include="xunit.runner.wpf" Version="$(xunitrunnerwpfVersion)" />
```
Comment (DustinCampbell) [inline build/ToolsetPackages/BaseToolset.csproj]: Ah! I guess this happens here now.

Comment (jaredpar) [reply build/ToolsetPackages/BaseToolset.csproj]: Yeah it had to move here. In the other position it was adding items to the reference graph and the SDK was unhappy. Here this is a simple net461 project for restoring tools so it fits more naturally.

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please

File: build/ToolsetPackages/BaseToolset.csproj (hunk 2 above)
Comment (DustinCampbell) [reply build/ToolsetPackages/BaseToolset.csproj]: Makes sense

File: build/Targets/Roslyn.Toolsets.Xunit.targets [hunk 3]
```diff
@@ -12,10 +12,39 @@
...
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <ForceGenerationOfBindingRedirects>true</ForceGenerationOfBindingRedirects>
+
+    <RuntimeIdentifier Condition="'$(TargetFramework)' == 'net461' AND '$(RuntimeIdentifier)' == '' AND '$(OS)' == 'Windows_NT'">win7</RuntimeIdentifier>
+  </PropertyGroup>
+    
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp2.0'">
+    <Content Include="$(NetStandard20RefPath)\netstandard.dll">
```
Comment (agocke) [inline build/Targets/Roslyn.Toolsets.Xunit.targets]: Last time I tried upgrading to the latest netcoreapp2.0 they included System.ValueTuple and tons of tests that were testing error scenarios around not having System.ValueTuple started failing. How is that not the case now?

Comment (jaredpar) [reply build/Targets/Roslyn.Toolsets.Xunit.targets]: The reason why is this is still the netstandard you were previously using for testing. Haven't changed this one yet. 

---
In reply to: [136669794](https://github.com/dotnet/roslyn/pull/21876#discussion_r136669794) [](ancestors = 136669794)

**Comment by alrz** [comment]
even though I've installed sdk (as mentioned in `Building, Debugging, and Testing on Windows.md`) running `restore.cmd` downloads a full copy to `Binaries\Tools\dotnet\sdk`. why is that?

**Comment by jaredpar** [comment]
> running restore.cmd downloads a full copy to Binaries\Tools\dotnet\sdk. is that expected?

Yes for now.

The Build.cmd / Restore.cmd files really just call into build\scripts\build.ps1 which is our official build script. That code is very particular about what version of tools it uses to build to ensure consistency in the official builds we produce. If it doesn't find tools at versions it deems compatible then it will download copies that are compatible and use those. That applies to the .NET Core SDK, MSBuild and NuGet.exe. 

The .NET Core SDK is still very new to us and we don't have a good feel for what compatibility will mean from build to build. Additionally we are relying on a couple of work arounds right now that aren't guaranteed to work on newer builds once the underlying bug is fixed. As such we look for an **exact** match. 

Other tools like MSBuild we have more lax versioning constraints versioning. Generally just look for a minimum version. I'm sure we'll move to that model for .NET Core SDK too but starting conservative for now.

File: build/Targets/Roslyn.Toolsets.Xunit.targets [hunk 4]
```diff
@@ -12,10 +12,39 @@
...
+      <Link>ref/%(FileName)%(Extension)</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <Visible>false</Visible>
+    </Content>
+    <Content Include="$(NetStandard20RefPath)\System.Dynamic.Runtime.dll">
+      <Link>ref/%(FileName)%(Extension)</Link>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <Visible>false</Visible>
+    </Content>
+    <Content Include="$(MicrosoftCSharpRefPath)\Microsoft.CSharp.dll">
```
Comment (jcouv) [inline build/Targets/Roslyn.Toolsets.Xunit.targets]: @nguerrera I found why the build copies the ref for Microsoft.CSharp.dll from nuget packages.

`build/Targets/Settings.props:    <MicrosoftCSharpRefPath>$(NuGetPackageRoot)/microsoft.csharp/$(MicrosoftCSharpVersion)/ref/netstandard1.0/</MicrosoftCSharpRefPath>`

Is the fallback folder for packages a new behavior of the 15.5/2.1 SDK?

Comment (jaredpar) [reply build/Targets/Roslyn.Toolsets.Xunit.targets]: The name "ref" here is probably a bit misleading. This has nothing to do with our actual build semantics. Rather it the directory that our core unit test function (compile, verify and run) look for reference assemblies in. We should probably rename this to "UnitTestRef" to make it much more apparent what the function of this folder is.

Comment (nguerrera) [reply build/Targets/Roslyn.Toolsets.Xunit.targets]: The fallback folder was in 2.0 as well. We'll have to dig in to understand why it wasn't populated/used before this change. There are supported ways to opt out (if need be), but I wouldn't expect any difference here between 2.0 and 2.1/15.5 so something unexpected is going on.

**Pull request merged**